File: filtrar matematica.py
# ============================================================
# Script Python - Filtrar dados de Matemática
# Gera um arquivo por TIPO DE ESCOLA
# ============================================================

import pandas as pd
import os
import re
from pathlib import Path

# ------------------------------------------------------------
# CONFIGURAÇÃO
# ------------------------------------------------------------

# Altere para o caminho onde estão os arquivos originais do PNLD
caminho = "/caminho/para/sua/pasta/PNLD"
filtro = "matem"

# Tipos de escola e padrões de busca nos nomes dos arquivos
TIPOS_ESCOLA = {
    "ESTADUAIS": "*ESTADUAIS*.xlsx",
    "FEDERAIS": "*FEDERAIS*.xlsx", 
    "MUNICIPAIS": "*MUNICIPAIS*.xlsx",
}

# ------------------------------------------------------------
# FUNÇÕES
# ------------------------------------------------------------

def extrair_ano(nome_arquivo):
    match = re.search(r'(\d{4})', nome_arquivo)
    return int(match.group(1)) if match else None

def processar_arquivo(arquivo, tipo_escola):
    """Processa um arquivo Excel com múltiplas abas"""
    dados = []
    
    try:
        todas_planilhas = pd.read_excel(arquivo, sheet_name=None)
        
        for nome_aba, df in todas_planilhas.items():
            # Converter nomes das colunas para string e limpar espaços
            df.columns = [str(c).strip() for c in df.columns]
            col_comp = [c for c in df.columns if "COMPONENTE" in c.upper()]
            
            if col_comp:
                # Converter coluna para string antes de filtrar
                coluna = df[col_comp[0]].fillna("").astype(str)
                df_filt = df[coluna.str.contains(filtro, case=False, na=False)]
                
                if len(df_filt) > 0:
                    df_filt = df_filt.copy()
                    df_filt["TIPO_ESCOLA"] = tipo_escola
                    df_filt["ANO"] = extrair_ano(arquivo.name)
                    df_filt["UF"] = nome_aba
                    dados.append(df_filt)
                    
    except Exception as e:
        print(f"    Erro: {e}")
    
    return dados

# ------------------------------------------------------------
# PROCESSAMENTO POR TIPO
# ------------------------------------------------------------

arquivos_gerados = []

for tipo, padrao in TIPOS_ESCOLA.items():
    print(f"\n{'='*50}")
    print(f"PROCESSANDO: {tipo}")
    print('='*50)
    
    arquivos = list(Path(caminho).glob(padrao))
    print(f"Arquivos encontrados: {len(arquivos)}")
    
    todos_dados = []
    
    for arquivo in arquivos:
        print(f"  → {arquivo.name}...", end=" ", flush=True)
        dados = processar_arquivo(arquivo, tipo.rstrip("S"))  # Remove "S" final
        todos_dados.extend(dados)
        total = sum(len(d) for d in dados)
        print(f"{total} registros")
    
    if todos_dados:
        df_tipo = pd.concat(todos_dados, ignore_index=True)
        
        # Se tiver mais de 1 milhão de linhas, salvar como CSV
        if len(df_tipo) > 1000000:
            arquivo_saida = os.path.join(caminho, f"MATEMATICA_{tipo}.csv")
            df_tipo.to_csv(arquivo_saida, index=False, encoding='utf-8-sig')
            formato = "CSV"
        else:
            arquivo_saida = os.path.join(caminho, f"MATEMATICA_{tipo}.xlsx")
            df_tipo.to_excel(arquivo_saida, index=False)
            formato = "XLSX"
        
        arquivos_gerados.append((tipo, arquivo_saida, len(df_tipo)))
        
        print(f"\n✓ Salvo: MATEMATICA_{tipo}.{formato.lower()} ({len(df_tipo):,} registros)")

# ------------------------------------------------------------
# RESUMO FINAL
# ------------------------------------------------------------

print(f"\n{'='*50}")
print("RESUMO FINAL")
print('='*50)

total_geral = 0
for tipo, arquivo, total in arquivos_gerados:
    print(f"  {tipo}: {total:,} registros")
    total_geral += total

print(f"\n  TOTAL GERAL: {total_geral:,} registros")
print(f"\nArquivos salvos em:\n{caminho}")

File: filtrar matematica.r
# ============================================================
# Script R - Filtrar dados de Matemática
# Gera um arquivo por TIPO DE ESCOLA
# ============================================================

# Instalar pacotes se necessário
pacotes <- c("openxlsx", "dplyr", "stringr")
for (p in pacotes) {
  if (!requireNamespace(p, quietly = TRUE)) {
    install.packages(p, repos = "https://cloud.r-project.org")
  }
}

library(openxlsx)
library(dplyr)
library(stringr)

# ------------------------------------------------------------
# CONFIGURAÇÃO
# ------------------------------------------------------------

# Altere para o caminho onde estão os arquivos originais do PNLD
caminho <- "/caminho/para/sua/pasta/PNLD"
filtro <- "matem"

# Tipos de escola e padrões de busca
tipos_escola <- list(
  ESTADUAIS = "ESTADUAIS.*\\.xlsx$",
  FEDERAIS = "FEDERAIS.*\\.xlsx$",
  MUNICIPAIS = "MUNICIPAIS.*\\.xlsm$"
)

# ------------------------------------------------------------
# FUNÇÕES
# ------------------------------------------------------------

extrair_ano <- function(nome_arquivo) {
  match <- str_extract(nome_arquivo, "\\d{4}")
  if (!is.na(match)) return(as.integer(match))
  return(NA)
}

processar_arquivo <- function(arquivo, tipo_escola) {
  cat("  ->", basename(arquivo), "... ")
  
  dados <- list()
  
  tryCatch({
    # Carregar workbook e obter nomes das abas
    wb <- loadWorkbook(arquivo)
    abas <- names(wb)
    
    for (aba in abas) {
      df <- read.xlsx(arquivo, sheet = aba)
      
      # Limpar nomes das colunas
      names(df) <- str_trim(names(df))
      
      # Encontrar coluna COMPONENTE
      col_comp <- names(df)[str_detect(toupper(names(df)), "COMPONENTE")]
      
      if (length(col_comp) > 0) {
        # Converter para character e filtrar por matemática
        df_filt <- df %>%
          mutate(across(all_of(col_comp[1]), as.character)) %>%
          filter(str_detect(tolower(.data[[col_comp[1]]]), filtro))
        
        if (nrow(df_filt) > 0) {
          df_filt <- df_filt %>%
            mutate(
              TIPO_ESCOLA = str_remove(tipo_escola, "S$"),
              ANO = extrair_ano(basename(arquivo)),
              UF = aba
            )
          dados <- append(dados, list(df_filt))
        }
      }
    }
    
    total <- sum(sapply(dados, nrow))
    cat(total, "registros\n")
    
  }, error = function(e) {
    cat("Erro:", e$message, "\n")
  })
  
  return(dados)
}

# ------------------------------------------------------------
# PROCESSAMENTO POR TIPO
# ------------------------------------------------------------

arquivos_gerados <- list()

for (tipo in names(tipos_escola)) {
  cat("\n", strrep("=", 50), "\n", sep = "")
  cat("PROCESSANDO:", tipo, "\n")
  cat(strrep("=", 50), "\n")
  
  padrao <- tipos_escola[[tipo]]
  arquivos <- list.files(caminho, pattern = padrao, full.names = TRUE)
  
  cat("Arquivos encontrados:", length(arquivos), "\n\n")
  
  todos_dados <- list()
  
  for (arquivo in arquivos) {
    dados <- processar_arquivo(arquivo, tipo)
    todos_dados <- append(todos_dados, dados)
  }
  
  if (length(todos_dados) > 0) {
    df_tipo <- bind_rows(todos_dados)
    
    # Se tiver mais de 1 milhão de linhas, salvar como CSV
    if (nrow(df_tipo) > 1000000) {
      arquivo_saida <- file.path(caminho, paste0("MATEMATICA_", tipo, ".csv"))
      write.csv(df_tipo, arquivo_saida, row.names = FALSE, fileEncoding = "UTF-8")
      formato <- "CSV"
    } else {
      arquivo_saida <- file.path(caminho, paste0("MATEMATICA_", tipo, ".xlsx"))
      write.xlsx(df_tipo, arquivo_saida, rowNames = FALSE)
      formato <- "XLSX"
    }
    
    arquivos_gerados[[tipo]] <- list(
      arquivo = arquivo_saida,
      registros = nrow(df_tipo),
      formato = formato
    )
    
    cat("\n Salvo: MATEMATICA_", tipo, ".", tolower(formato), 
        " (", format(nrow(df_tipo), big.mark = ","), " registros)\n", sep = "")
  }
}

# ------------------------------------------------------------
# RESUMO FINAL
# ------------------------------------------------------------

cat("\n", strrep("=", 50), "\n", sep = "")
cat("RESUMO FINAL\n")
cat(strrep("=", 50), "\n")

total_geral <- 0
for (tipo in names(arquivos_gerados)) {
  info <- arquivos_gerados[[tipo]]
  cat(" ", tipo, ": ", format(info$registros, big.mark = ","), " registros (", info$formato, ")\n", sep = "")
  total_geral <- total_geral + info$registros
}

cat("\n  TOTAL GERAL: ", format(total_geral, big.mark = ","), " registros\n", sep = "")
cat("\nArquivos salvos em:\n", caminho, "\n")
